### PR TITLE
Replace return_address usage for rooting with stack guards and convenience macros.

### LINKED ISSUE #34227: return_address intrinsic problematic with MIR and hard/impossible to use safely.
Servo currently can't compile with `-Zorbit` (see servo/servo#11706) and my suspect is `return_address`.

Even if that wasn't the reason, Servo's usage of `return_address` to create a rooting cycle only ever worked because it happened to get the final destination of the call, but with MIR, the call destination might be a temporary, so it's not the final destination anymore.
Servo has a lint for disallowing moves on the types that are being returned, but that analysis is wrong unless performed on *exactly* the MIR we translate.

We also could change the ABI at any time, so the indirect returns becomes direct (we can do this for any pair of pointer-sized values) which would also break Servo.
I'm not even sure how LLVM doesn't miscompile what Servo is doing, capturing an `sret` argument seems quite dubious to me, and some architectures also have an intermediary copy.

AFAIK, the eventual solution is to have proper stack map support, however that might be a long ways off and I don't want to block removing old trans on it.

cc @jdm @pcwalton @Manishearth @rust-lang/compiler

### PR DESCRIPTION
**Pull request by eddyb**
The existing `Rooted` and `RootedVec` users were migrated the the following two macros:

``` rust
let x = Rooted::new(cx, value);
// Was changed to:
rooted!(in(cx) let x = value);
// Which expands to:
let mut __root = Rooted::new_unrooted(value);
let x = RootedGuard::new(cx, &mut __root);
```

``` rust
let mut v = RootedVec::new();
v.extend(iterator);
// Was changed to:
rooted_vec!(let v <- iterator);
// Which expands to:
let mut __root = RootableVec::new();
let v = RootedVec::new(&mut __root, iterator);
```

The `rooted!` macro depends on servo/rust-mozjs#272.
These APIs based on two types, a container to be rooted and a rooting guard, allow implementing both `Rooted`-style rooting and `Traceable`-based rooting in stable Rust, without abusing `return_address`.

Such macros may have been tried before, but in 1.9 their hygiene is broken, they work only since 1.10.

Sadly, `Rooted` is a FFI type and completely exposed, so I cannot prevent anyone from creating their own, although all fields but the value get overwritten by `RootedGuard::new` anyway.
`RootableVec` OTOH is _guaranteed_ to be empty when not rooted, which makes it harmless AFAICT.

By fixing rust-lang/rust#34227, this PR enables Servo to build with `-Zorbit`.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix rust-lang/rust#34227
- [x] These changes do not require tests because they are not functional changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11872)

### CODE REVIEW

File: components/script/dom/macros.rs [hunk 1]
```diff
@@ -503,3 +503,15 @@ macro_rules! document_and_element_event_handlers(
         event_handler!(paste, GetOnpaste, SetOnpaste);
     )
 );
+
+#[macro_export]
+macro_rules! rooted_vec {
+    (let $name:ident = $iter:expr) => {
+        let mut __root = $crate::dom::bindings::trace::RootableVec::new_unrooted();
```
Comment (Manishearth) [inline components/script/dom/macros.rs]: make these unsafe and make the macro safe?

Comment (Manishearth) [inline components/script/dom/macros.rs]: oh I see what you did there

**Comment by Manishearth** [comment]
r? @jdm 

The design for RootedVec seems sound to me (will look at `rooted!` later), but I'm not as familiar with the intricacies of these bindings as jdm is

**Comment by Manishearth** [comment]
```
   Compiling glutin_app v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/ports/glutin)
/home/servo/buildbot/slave/linux-dev/build/components/script/webdriver_handlers.rs:31:41: 31:55 error: unresolved import `js::jsapi::UndefinedValue`. There is no `UndefinedValue` in `js::jsapi` [E0432]
/home/servo/buildbot/slave/linux-dev/build/components/script/webdriver_handlers.rs:31 use js::jsapi::{JSContext, HandleValue, UndefinedValue};
                                                                                                                              ^~~~~~~~~~~~~~
```

**Comment by eddyb** [comment]
I think I have a fix for `/_mozilla/mozilla/sequence-hole.html` - the `Vec<T>` `FromJSConvertible` in `jsapi::conversions` wasn't unrooting properly.
Is `/html/semantics/embedded-content/the-iframe-element/change_parentage.html` a legitimate failure?

FWIW, those are the only failures I've seen with `-Zorbit` on, which is probably good news!

**Comment by jdm** [comment]
change_parentage is #11703.

**Comment by eddyb** [comment]
@jdm The convertible failure is fixed FWIW, I just checked (with `-Zorbit` no less!). Also, although dromaero is ridiculously variable, switching to MIR seems to universally drop the times down, so that's something (again, if it's not completely random variance). Someone else should look into it.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 2]
```diff
@@ -4475,13 +4474,13 @@ def getBody(self):
             attrs = "JSPROP_ENUMERATE"
             if self.descriptor.operations['IndexedSetter'] is None:
                 attrs += " | JSPROP_READONLY"
-            fillDescriptor = ("desc.get().value = result_root.ptr;\n"
```
Comment (Ms2ger) [inline components/script/dom/bindings/codegen/CodegenRust.py]: This is a behaviour change, is it not? Can we avoid those in this PR? Keeping the buggy code with a TODO comment until such time we can write a test for it seems better to me.

**Comment by Ms2ger** [comment]
SGTM except for the behaviour changes.

CC @nox @Manishearth

**Comment by eddyb** [comment]
@Ms2ger Updated with the behavior changes removed and the proper rust-mozjs commit set.

**Comment by nox** [comment]
SGTM too.

@jdm I've been told by @eddyb that you don't wish to see that merged. Why? We can't progress on the MIR front without it.

**Comment by eddyb** [comment]
Maybe I was a bit unclear on IRC: my understanding was that @jdm wanted to wait until the excitement around Servo nightlies tapers off before risking to merge such a core change.
Although IMO the risk is not that high, compared to say, touching anything to do with `JS<T>` or `Root<T>`.

**Comment by jdm** [comment]
I intend to review this next week.

File: components/script/dom/macros.rs [hunk 3]
```diff
@@ -503,3 +503,18 @@ macro_rules! document_and_element_event_handlers(
         event_handler!(paste, GetOnpaste, SetOnpaste);
     )
 );
+
+#[macro_export]
+macro_rules! rooted_vec {
+    (let mut $name:ident) => {
+        rooted_vec!(let mut $name <- None.into_iter())
```
Comment (Ms2ger) [inline components/script/dom/macros.rs]: `iter::empty()`?

**Comment by Ms2ger** [comment]
One nit, and this LGTM. Will wait for jdm to wake up, though.

File: components/script/dom/macros.rs (hunk 3 above)
Comment (eddyb) [inline components/script/dom/macros.rs]: Ah that is indeed a thing, thanks!

**Comment by Ms2ger** [comment]
And fix the unused `std::iter::IntoIterator` import in `trace.rs`.

File: components/script/dom/bindings/proxyhandler.rs [hunk 4]
```diff
@@ -30,18 +30,19 @@ static JSPROXYSLOT_EXPANDO: u32 = 0;
 pub unsafe extern "C" fn get_property_descriptor(cx: *mut JSContext,
                                                  proxy: HandleObject,
                                                  id: HandleId,
-                                                 desc: MutableHandle<PropertyDescriptor>)
+                                                 mut desc: MutableHandle<PropertyDescriptor>)
```
Comment (Ms2ger) [inline components/script/dom/bindings/proxyhandler.rs]: Revert

File: components/script/dom/document.rs [hunk 5]
```diff
@@ -1009,18 +1009,15 @@ impl Document {
             }
         }
 
-        let mut touches = RootedVec::new();
+        rooted_vec!(let mut touches);
         touches.extend(self.active_touch_points.borrow().iter().cloned());
```
Comment (jdm) [inline components/script/dom/document.rs]: Shouldn't this use the <- syntax?

**Comment by jdm** [comment]
A+ work. Thanks for doing this, @eddyb!

**Comment by eddyb** [comment]
@Ms2ger Travis passes with warnings?

**Pull request merged**